### PR TITLE
fix: redirect /hub/insights/new to login if not authenticated

### PR DESCRIPTION
**Pull request by brandonroberts**
## Description

- Redirects `/hub/insights/new` to `/signin` to prompt user to login if not authenticated
- After authenticating/connecting, the user will be redirected to the new repository insight page

## What type of PR is this? (check all applicable)

- [ ] 🍕 Feature
- [x] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI
- [ ] 📦 Chore (Release)
- [ ] ⏩ Revert

## Related Tickets & Documents

Fixes https://github.com/open-sauced/app/issues/2599

## Mobile & Desktop Screenshots/Recordings

https://github.com/open-sauced/app/assets/42211/9d70be0e-305d-4afa-b525-434207f4e43c

## Steps to QA

1. Logout
2. Navigate to `/hub/insights/new`
3. Redirected to `/workspaces/new`
4. Note the `Create Workspace` button is covered with `Connect to GitHub`
5. Enter a name and description for the workspace
6. Add some tracked repositories
7. Click `Connect to GitHub`
8. After logging in, you're navigated back to the new workspace page with the information filled out
9. Click `Create Workspace`

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 docs.opensauced.pizza
- [ ] 🍕 dev.to/opensauced
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

### CODE REVIEW

**Comment by bdougie** [comment]
The auth flow works, but I land on the highlights feed. TBH, the highlights feed is deprioritized and per the .new requirements we need the user to redirect back to `/hub/insights/new` after authenticating.

**Review by jpmcb — APPROVED** [review]

**Comment by brandonroberts** [comment]
> The auth flow works, but I land on the highlights feed. TBH, the highlights feed is deprioritized and per the .new requirements we need the user to redirect back to `/hub/insights/new` after authenticating.

This is ready to go now and will redirect to the new repository insight page for their personal after being prompted to login.


https://github.com/open-sauced/app/assets/42211/c534c9af-dd20-481e-9ecc-0b33cf79fb3a

**Comment by nickytonline** [comment]
> > The auth flow works, but I land on the highlights feed. TBH, the highlights feed is deprioritized and per the .new requirements we need the user to redirect back to `/hub/insights/new` after authenticating.
> 
> This is ready to go now and will redirect to the new repository insight page for their personal after being prompted to login.
> 
>  Screen.Recording.2024-03-13.at.10.33.13.AM.mov

Can we redirect server-side instead back to the new insight page, or is there a limitation with how we log in with GH/Supabase? Currently it goes back to the feed and then redirects what appears to be client-side.

**Comment by bdougie** [comment]
> Can we redirect server-side instead back to the new insight page, or is there a limitation with how we log in with GH/Supabase? Currently it goes back to the feed and then redirects what appears to be client-side.

Was curious why the redirect back to the feed and then to /new. I assume this is why.

**Comment by brandonroberts** [comment]
> > > The auth flow works, but I land on the highlights feed. TBH, the highlights feed is deprioritized and per the .new requirements we need the user to redirect back to `/hub/insights/new` after authenticating.
> > 
> > 
> > This is ready to go now and will redirect to the new repository insight page for their personal after being prompted to login.
> > Screen.Recording.2024-03-13.at.10.33.13.AM.mov
> 
> Can we redirect server-side instead back to the new insight page, or is there a limitation with how we log in with GH/Supabase? Currently it goes back to the feed and then redirects what appears to be client-side.

Yea, its two-fold from what I came across.

Supabase doesn't provide an API to generate a login URL server-side, and building the URL manually didn't seem like a good option.

Also, requesting the user session immediately after logging doesn't return a valid session in the middleware, so it falls back to the login page in a loop. Letting the request fall back to the client creates the user session and triggers the redirect correctly.

**Comment by brandonroberts** [comment]
We can navigate back to the `/signin` page and redirect from there if its less jarring

**Comment by brandonroberts** [comment]
It now redirects back to `/signin` after logging in before going to its destination path.

File: pages/signin.tsx [hunk 1]
```diff
@@ -1,16 +1,43 @@
...
+    destination?: string;
+    auth?: string;
   };
+  const hasParams = Object.keys(router.query).length > 0;
+  const hasAuth = auth === "true";
+
+  useEffect(() => {
+    if (hasParams) {
+      if (destination && hasAuth) {
+        router.replace(decodeURIComponent(destination));
```
Comment (nickytonline) [inline pages/signin.tsx:26]: refactor (blocking): `router.query` already decodes the querystring values destructured above. Same for the `redirectedFrom` below.

```suggestion
router.replace(destination);
```

**Review by nickytonline — CHANGES_REQUESTED** [review]

**Comment by bdougie** [comment]
> * /hub/insights/new

I get an infinite redirect now.

Rewatch GIF Recording - 2024-03-13 at 5 01 21 PM

**Comment by brandonroberts** [comment]
Redirect loop issues should be resolved now. Tested on the deploy preview

**Comment by bdougie** [comment]
I am still redirecting non-stop.

Screen Shot 2024-03-15 at 11 20 31 AM

**Comment by bdougie** [comment]
What is we approach this differently and allow users to see the workspace creation without logging in and force a login and they hit create.

**Comment by brandonroberts** [comment]
> What if we approach this differently and allow users to see the workspace creation without logging in and force a login and they hit create.

That seems reasonable to me. We already have some login there for using `Create New Workspace` from the dropdown

**Comment by brandonroberts** [comment]
@bdougie This is ready for another review. `/hub/insights/new` now redirects to creating a new workspace without being logged in and keeps the information entered after logging in.

**Review by bdougie — APPROVED** [review]
works as expected.

**Pull request merged**